foolbox/attacks/adef_attack.py

```diff
@@ -0,0 +1,405 @@
+    overshoot : float >= 1
+        Multiply the resulting vector field by this number,
+        if deformed image is still correctly classified
+        (default is overshoot = 1 for no overshooting).
+    smooth : float >= 0
+        Width of the Gaussian kernel used for smoothing.
+        (default is smooth = 0 for no smoothing).
+    targeting : bool
+        targeting = False (default) to stop as soon as model misclassifies
+        input targeting = True to stop only once a candidate label is achieved.
+
+    """
+
+    def __init__(self, model=None, criterion=Misclassification()):
+        super(ADefAttack, self).__init__(model=model, criterion=criterion)
+        self.vector_field = None
+
+    @call_decorator
+    def __call__(self, input_or_adv, ind_of_candidates=1, unpack=True,
+                 max_iter=100, max_norm=np.inf, label=None,
+                 overshoot=1.1, smooth=1.0, targeting=False):
+
+        a = input_or_adv
+        del input_or_adv
+        del label
+        del unpack
+
+        if not a.has_gradient():
+            return
+
+        # Include the correct label (index 0) in the list of targets.
+        # Remove duplicates and sort the label indices.
+        ind_of_candidates = np.unique(np.append(ind_of_candidates, 0))
+        # Remove negative entries.
+        ind_of_candidates = ind_of_candidates[ind_of_candidates >= 0]
+
+        # Number of classes to target + 1 ( >= 2).
+        # Example 1: num_classes = 10 for MNIST
+        # Example 2: If ind_of_candidates == 1,
+        # then only the second highest label is targeted
+        num_classes = ind_of_candidates.size
+
+        n = 0  # iteration number
+
+        perturbed = a.original_image.copy()  # is updated in every iteration
+
+        image_original = a.original_image.copy()  # is not updated,
+        # but kept as a copy
+
+        color_axis = a.channel_axis(batch=False)  # get color axis
+        assert color_axis in [0, 2]
+        hw = [perturbed.shape[i] for i in range(perturbed.ndim)
+              if i != color_axis]
+        h, w = hw
+
+        logits, grad, is_adv = a.predictions_and_gradient(perturbed)
+
+        logits = np.expand_dims(logits, axis=0)
+
+        # Indices of the 'num_classes' highest values in descending order:
+        candidates = np.argsort(-logits)[:, ind_of_candidates]
+        original_label = candidates[:, 0]
+
+        # fx[lab] is negative if the model prefers the original label
+        # for x over the label 'lab'.
+        fx = (logits.transpose() - logits[0, original_label]).transpose()
+
+        norm_full = 0  # norm of the vector field
+        vec_field_full = np.zeros((h, w, 2))  # the vector field
+
+        current_label = original_label
+        logging.info('Iterations finished: 0')
+        logging.info('Current labels: {} '.format(current_label))
+
+        for step in range(max_iter):
+            n += 1
+            logits, grad, is_adv = a.predictions_and_gradient(perturbed)
```

Comment:
evaluating the gradient here seems unnecessary, now that backward is called later and the returned grad is never used as far as I see